packages/camera/camera_avfoundation/ios/camera_avfoundation/Sources/camera_avfoundation/FLTCam.m

```diff
@@ -1331,6 +1365,19 @@ - (void)setUpCaptureSessionForAudio {
   // Setup the audio output.
   _audioOutput = [[AVCaptureAudioDataOutput alloc] init];
 
+  dispatch_block_t block = ^{
+    // Setup options implicit to AVAudioSessionCategoryPlayback to not disturb video_player.
+    upgradeAudioSessionCategory(AVAudioSessionCategoryPlayAndRecord,
+                                AVAudioSessionCategoryOptionDefaultToSpeaker |
+                                    AVAudioSessionCategoryOptionAllowBluetoothA2DP |
+                                    AVAudioSessionCategoryOptionAllowAirPlay);
+  };
+  if (!NSThread.isMainThread) {
+    dispatch_sync(dispatch_get_main_queue(), block);
+  } else {
+    block();
```

Comment:
is `setUpCaptureSessionForAudioIfNeeded` already guaranteed to be on background? Can you double check the caller of this function? If it's already on background, we can simply do 
```
NSAssert(!NSThread.isMainThread);
dispatch_sync(dispatch_get_main_queue()) {
  // your logic here
};

Reply:
It is called on the main thread in tests if I remember correctly https://github.com/flutter/packages/pull/7143#discussion_r1723565703.

Reply:
is it only called on main thread in tests? Can you give an example of such failure? I'd like to see if it's possible to run on background thread in those tests

Reply:
Try to run with this:

``` objc
  //if (!NSThread.isMainThread) {
    dispatch_sync(dispatch_get_main_queue(), block);
  //} else {
    //block();
  //}
```

And you will get this inside xcode:

```
Thread 1: EXC_BREAKPOINT (code=1, subcode=0x18082dc74)
```

When running outside of xcode with debugging there will be probably just crash, you can add `NSLog(@"%@", NSThread.currentThread);` to see that it runs on main thread `<_NSMainThread: 0x2804b0080>{number = 1, name = main}`.

You will probably need to run `flutter pub upgrade` first as without that I am getting this error during `flutter run` (win32 on macos?):

```
Could not build the precompiled application for the device.
Error (Xcode): ../../../../../../../../.pub-cache/hosted/pub.dev/win32-5.2.0/lib/src/guid.dart:32:9: Error: Type 'UnmodifiableUint8ListView' not found.
```

Reply:
> It is called on the main thread in tests if I remember correctly https://github.com/flutter/packages/pull/7143#discussion_r1723565703.

> When running outside of xcode with debugging there will be probably just crash, you can add NSLog(@"%@", NSThread.currentThread); to see that it runs on main thread <_NSMainThread: 0x2804b0080>{number = 1, name = main}.

So  `setUpCaptureSessionForAudioIfNeeded` is called on main thread somewhere (not just in test). Could you find the place it's called on main? The session setup shouldn't happen on main thread.

Reply:
No I did not mean that I observed something like that. I wrote that because I think that native tests can be run also outside of xcode.

Reply:
hmmm im confused, maybe im missing something. `setUpCaptureSessionForAudioIfNeeded` should only be called on background thread, so the same applies when testing this function, it should only be tested on background thread as well.

Reply:
Seems `FLTCam` tries to act like something independent and almost unaware of what queue it actually should run on and dispatches to the session queue are made in `CameraPlugin`. Tests are testing `FLTCam` not `CameraPlugin`. There are for example many calls to `startVideoRecordingWithCompletion` from tests. Now for example `reportErrorMessage` is called only from functions which are always called on session queue (outside of tests) so it itself is called on only on session queue but it uses `FLTEnsureToRunOnMainQueue` which tests whether we are on main queue or not, this suggests that it is probably intended that tests use `FLTCam` on main thread.

Reply:
> but it uses FLTEnsureToRunOnMainQueue which tests whether we are on main queue or not, this suggests that it is probably intended that tests use FLTCam on main thread.

No, `FLTEnsureToRunOnMainQueue` doesn't test whether we are on main queue or not. Instead, it dispatches to main queue if it's not already on main: https://github.com/flutter/packages/blob/45bcc3210fcc24011bc7f88de1ca61528d9e5cbc/packages/camera/camera_avfoundation/ios/camera_avfoundation/Sources/camera_avfoundation/QueueUtils.m

Reply:
By "it tests" I meant this: `if (!NSThread.isMainThread) {`

Reply:
Gotcha I misunderstood you. 

Right now we have the tests and production code calling separate paths: 
```
if background {
  // production code runs this
} else {
  // test code runs this
}
```

Is it hard to modify the tests to ensure session is setup on background? Something like: 

```

testFoo() {
  let expectation = ...
  sessionQueue.async {
    test here
    expectation.fulfill()
  }
  waitForExpectation()
}
```

Reply:
This work is already done, in `CameraPlugin`, all that would be needed to do is to call functions from this class instead of `FLTCam`. But I do not think that there is any problem. I suppose these tests are supposed to be unit tests for `FLTCam` class and there is mostly not even used real (not mocked) `AVFoundation`.

Reply:
> This work is already done, in CameraPlugin, all that would be needed to do is to call functions from this class instead of FLTCam. 

What work is already done?

Reply:
About dispatching on the session queue.

Reply:
Sorry im very confused. Does the test already dispatch to session queue? If it does, why are you still getting the crash? Otherwise, we should do it, so the test can pass.

Reply:
It is not just a single test or my test, I guess most of the tests do that, I saw some 7 calls to just `startVideoRecordingWithCompletion`.

Reply:
`CameraPlugin` and `FLTCam` are different classes and each has their own tests. Is the `FLTCam` the one fails? If so, we should dispatch to the background thread in the tests. 

The problem is that right now you have a code branch that is only run by tests but not in production, which defeats the purpose of tests.

Reply:
Btw I do not see a single native test calling for example `startVideoRecordingWithStreaming`.

Reply:
It also branches differently in most tests than in production.

Reply:
So if there is a problem let's do it similarly, I will add the function `FLTEnsureToRunOnMainQueueSync` which will be used here.

Reply:
> It also branches differently in most tests than in production.

I haven't touched this plugin for a while. Can you point me to the code? IIRC both of the branches are called in production.

Reply:
> I haven't touched this plugin for a while. Can you point me to the code? IIRC both of the branches are called in production.

But not on all places where it is used like I already mentioned earlier about `reportErrorMessage`.

> The problem is that right now you have a code branch that is only run by tests but not in production, which defeats the purpose of tests.

I suppose you meant that in reverse that the problem is that there is a branch which is called in production but not in tests, tests should cover what is in production not vice versa. And btw this is not entirely true, there is an integration test which also calls that second branch on the session queue so both are actually covered already.

Reply:
> I suppose you meant that in reverse that the problem is that there is a branch which is called in production but not in tests

No. 
In this code: 

```
  if (!NSThread.isMainThread) {
    dispatch_sync(dispatch_get_main_queue(), block);
  } else {
    block(); 
  }
```
The `if` branch is only run in production, and `else` is only run in test right? Or am i mistaken?

Reply:
Both are run in tests. In production only one of them.

Reply:
Oh sorry I confused `camera` with `video_player`. I am not sure whether there is such a test which tests with a real camera device but even if yes this has some limitations of course, like it probably cannot be tested automatically very well. So yes basically you are right. So what about that sync version of `FLTEnsureToRunOnMainQueue`?

Reply:
Or maybe I can just simply modify my own test to use a background queue and then both branches will be covered.

Reply:
> So what about that sync version of FLTEnsureToRunOnMainQueue?

This would be the same problem right? I think the right approach is the change the test, so that it dispatch to the background thread. The non-test code itself should assert that it's on the background thread.

Reply:
> This would be the same problem right?

Actually not, a separate function can be separately tested like is done with that one.

> The non-test code itself should assert that it's on the background thread.

This is actually by "definition".

Reply:
> Actually not, a separate function can be separately tested like is done with that one.

The problem here is not just the test coverage, but it's that you have a branch in code that is only called by tests. 

> This is actually by "definition".

What do you mean? To be clear, I was suggesting: 

```
NSAssert(!NSThread.isMainThread);
dispatch_sync(dispatch_get_main_queue()) {
  // your logic here
};
```

Reply:
But that same situation is also with `FLTEnsureToRunOnMainQueue`. It could be removed in several places where it has meaning only for tests. And if this is somehow fine because that function is maybe also used on places where it may run on the main thread in production then I have probably bad luck that I need sync dispatch instead of async. I think it is possible to issue `dispatch_async` and then wait until it finishes but that would feel somewhat weird. This looks to me more like an artificial problem. And I do not think that refactoring most of the tests is in scope of this PR which is intended to fix a bug.

> The problem here is not just the test coverage, but it's that you have a branch in code that is only called by tests.

How exactly is this a problem anyway? There are many other "adaptations" only because of tests, like interfaces and factories just to simplify testing, in both `camera` and `video_player`.

Reply:
> But that same situation is also with FLTEnsureToRunOnMainQueue.

Could you explain the problem with FLTEnsureToRunOnMainQueue? Maybe a concrete example? 

> How exactly is this a problem anyway? There are many other "adaptations" only for tests, like interfaces and factories just to simplify testing, in both camera and video_player.

I don't know why you are comparing with interfaces and factories. 

Imagine you have this code: 

```
func setupSession() {
  if is background thread {
    Logic A
  } else {
    Logic B
  }
}
```
Here your production code only calls Logic A, and test code only calls Logic B. Then the test isn't really providing any confidence right? 

What I suggest is: 

```
func setupSession() { 
  Assert(session must be setup on background);
  Logic A;
}
```

Then in the test, you can do: 

```
func testSetupSession() {
  dispatch_to_background {
    setupSession();
  }
}
```
This way, the test code actually covers Logic A, providing us with confidence in production.

Reply:
Your assertion will fail because other tests call `setupSession` on the main thread. And would this not cause a deadlock due to `dispatch_sync` waiting for block dispatched on main queue while also `waitForExpectation` waiting on main thread until `dispatch_sync` is done and expectation fulfilled? https://github.com/flutter/packages/pull/7143/#discussion_r1876564910

```
testFoo() {
  let expectation = ...
  sessionQueue.async {
    dispatch_sync(main_queue(), ...)
    expectation.fulfill()
  }
  waitForExpectation()
}
```

What I proposed does not have these problems and tests also cover production logic as I wrote.

Reply:
> And would this not cause a deadlock due to dispatch_sync waiting for block dispatched on main queue while also waitForExpectation is waiting on main thread until dispatch_sync is done and expectation fulfilled?

Oh I overlooked the part that it's `dispatch_sync` to main, which leads to deadlock in test. 

> What I proposed does not have such problems and tests also cover production logic as I wrote.

It would be the same problem, that `dispatch_sync(main)` is inherently not testable - You simply moved the same problem to `FLTEnsureToRunOnMainQueueSync` (i.e. the `dispatch_sync` branch of this helper would not be testable for the same reason). So I'd accept the current code as it is.

Reply:
> It would be the same problem

Oh I did not think of that, but as you wrote that, I realized that means that also the async test should be untestable, maybe it actually waits in some non blocking way. But anyway I did not see that documented anywhere, maybe that async test is not 100% clean then?

Reply:
> I realized that means that also the async test should be untestable

`dispatch_async` is different and it can be tested. The problem with `dispatch_sync` is that it causes deadlock when dispatching to the same queue.